sw/host/opentitanlib/src/spiflash/sfdp.rs

```diff
@@ -0,0 +1,529 @@
+    field!(wait_states_112 -> u8, 3, 16, 5);
+    field!(mode_bits_112 -> u8, 3, 21, 3);
+    field!(read_opcode_112 -> u8, 3, 22, 8);
+
+    field!(support_fast_read_222 -> bool, 4, 0, 1);
+    field!(support_fast_read_444 -> bool, 4, 4, 1);
+
+    field!(wait_states_222 -> u8, 5, 16, 5);
+    field!(mode_bits_222 -> u8, 5, 21, 3);
+    field!(read_opcode_222 -> u8, 5, 22, 8);
+    field!(wait_states_444 -> u8, 6, 16, 5);
+    field!(mode_bits_444 -> u8, 6, 21, 3);
+    field!(read_opcode_444 -> u8, 6, 22, 8);
+
+    field!(sector_type1_size -> u8, 7, 0, 8);
+    field!(sector_type1_opcode -> u8, 7, 8, 8);
+    field!(sector_type2_size -> u8, 7, 16, 8);
+    field!(sector_type2_opcode -> u8, 7, 24, 8);
+    field!(sector_type3_size -> u8, 8, 0, 8);
+    field!(sector_type3_opcode -> u8, 8, 8, 8);
+    field!(sector_type4_size -> u8, 8, 16, 8);
+    field!(sector_type4_opcode -> u8, 8, 24, 8);
+}
+
+/// `BlockEraseSize` represents whether or not the device can perform
+/// a 4KiB erase.
+#[derive(Debug, Eq, PartialEq, FromPrimitive, Clone, Copy, Serialize)]
+#[repr(u32)]
+pub enum BlockEraseSize {
+    Reserved0 = 0,
+    Block4KiB = 1,
+    Reserved2 = 2,
+    BlockNot4KiB = 3,
+    #[num_enum(default)]
+    Invalid,
+}
+
+impl Default for BlockEraseSize {
+    fn default() -> Self {
+        BlockEraseSize::Invalid
+    }
+}
+
+/// `WriteGranularity` represents whether or not the device has an internal
+/// buffer for program operations.
+#[derive(Debug, Eq, PartialEq, FromPrimitive, Clone, Copy, Serialize)]
+#[repr(u32)]
+pub enum WriteGranularity {
+    Granularity1Byte = 0,
+    Granularity64Bytes = 1,
+    #[num_enum(default)]
+    Invalid,
+}
+
+impl Default for WriteGranularity {
+    fn default() -> Self {
+        WriteGranularity::Invalid
+    }
+}
+
+/// `SupportedAddressModes` represents which addressing modes are valid for
+/// the device.
+#[derive(Debug, Eq, PartialEq, FromPrimitive, Clone, Copy, Serialize)]
+#[repr(u32)]
+pub enum SupportedAddressModes {
+    Mode3b = 0,
+    Mode3b4b = 1,
+    Mode4b = 2,
+    Reserved = 3,
+    #[num_enum(default)]
+    Invalid,
+}
+
+impl Default for SupportedAddressModes {
+    fn default() -> Self {
+        SupportedAddressModes::Invalid
+    }
+}
+
+/// `FastReadParam` represents the parameters for the different styles of fast read.
+#[derive(Default, Debug, Serialize)]
+pub struct FastReadParam {
+    pub wait_states: u8,
+    pub mode_bits: u8,
+    pub opcode: u8,
+}
+
+/// `Sector` represents the supported erase sector sizes of the device.
+#[derive(Default, Debug, Serialize)]
+pub struct Sector {
+    pub size: u32,
+    pub erase_opcode: u8,
+}
+
+/// The JEDEC parameter table is documented in JESD216 "Serial Flash Discoverable Parameters".
+/// This structure represents the parameter table after decoding it from the packed bitfield
+/// representation documented by JEDEC.
+#[derive(Default, Debug, Serialize)]
+pub struct JedecParams {
+    // Erase granularity.
```

Comment:
These aren't doc comments. Maybe worth fixing that?

Reply:
Done.